api/v1_projects.inc

```diff
@@ -225,40 +226,81 @@ function api_v1_project_pagedetails($method, $data, $query_params)
+            SELECT
+                image,
+                %s AS text,
+                %s AS user,
+                state
+            FROM %s
+            WHERE image = '%s'
+        ", $text_column, $user_column, $project->projectid,
+            DPDatabase::escape($data[":pagename"])
+        );
+        $result = DPDatabase::query($sql);
+        $row = mysqli_fetch_assoc($result);
+        $row["image_url"] = $project->url . "/" . $row["image"];
+
+        // We can't show the username here unless the user has proofed the page
+        // in one round or they are an PF/SA. We need to abstract the conditional
+        // logic out of tools/project_manager/page_detail.inc
+        unset($row["user"]);
+
+        return render_project_page_json($row);
     } else {
-        $round = get_Round_for_round_id($data[":pageroundid"]);
-        $text_column = $round->text_column_name;
-        $user_column = $round->user_column_name;
-    }
+        // PUT
+        global $pguser;
 
-    $sql = sprintf("
-        SELECT
-            image,
-            %s AS text,
-            %s AS user,
-            state
-        FROM %s
-        WHERE image = '%s'
-    ", $text_column, $user_column, $data[":projectid"]->projectid,
-        DPDatabase::escape($data[":pagename"])
-    );
-    $result = DPDatabase::query($sql);
-    $row = mysqli_fetch_assoc($result);
-    $row["image_url"] = $data[":projectid"]->url . "/" . $row["image"];
-
-    // We can't show the username here unless the user has proofed the page
-    // in one round or they are an PF/SA. We need to abstract the conditional
-    // logic out of tools/project_manager/page_detail.inc
-    unset($row["user"]);
-
-    return render_project_page_json($row);
+        if ($data[":pageroundid"] == "OCR") {
+            throw new BadRequest("It is not possible to work in round OCR");
+        } else {
+            validate_user_can_work_in_project($project, $data[":pageroundid"]);
+        }
+        try {
+            $request_body = api_get_request_body();
+            $old_page_state = $request_body->old_page_state;
+            $new_page_state = $request_body->new_page_state;
+            $lpage = get_indicated_LPage($project->projectid, $project->state, $data[":pagename"], $old_page_state, 0);
+            switch ($new_page_state) {
+            case "available":
+                $lpage->returnToRound($pguser);
+                return [];
+                break;
+            case "done":
+                [$page_saved, $limit_reached] = $lpage->attemptSaveAsDone($request_body->text, $pguser);
+                return ["pageSaved" => $page_saved, "limitReached" => $limit_reached];
+                break;
+            case "resume":
+                $lpage->resume_saved_page($pguser);
+                return render_page_data($lpage);
+                break;
+            case "temp":
+                $lpage->saveAsInProgress($request_body->text, $pguser);
+                if ($request_body->revert) {
```

Comment:
Are you saying that there shouldn't be a way to get page text from the previous round in the API to do a "restore to original" or that we should not use LPage for this purpose?
The function used allows for previous skipped rounds.

Reply:
I was thinking about not using the LPage for it. However, I had totally forgotten about skipped rounds (because UGH). In that case I think we need to keep the revert capability via the API. I don't want to force the client to encode that logic.